pkg/coscheduling/coscheduling.go

```diff
@@ -95,295 +80,316 @@ func New(obj runtime.Object, handle framework.FrameworkHandle) (framework.Plugin
+func (cs *Coscheduling) Less(podInfo1, podInfo2 *framework.QueuedPodInfo) bool {
+	prio1 := podutil.GetPodPriority(podInfo1.Pod)
+	prio2 := podutil.GetPodPriority(podInfo2.Pod)
+	if prio1 != prio2 {
+		return prio1 > prio2
 	}
-
-	// If it's not a regular Pod, store the PodGroup in PodGroupInfos
-	if len(pgKey) > 0 {
-		cs.podGroupInfos.Store(pgKey, pgInfo)
+	creationTime1 := cs.pgMgr.GetCreationTime(podInfo1.Pod, podInfo1.InitialAttemptTimestamp)
+	creationTime2 := cs.pgMgr.GetCreationTime(podInfo2.Pod, podInfo2.InitialAttemptTimestamp)
+	if creationTime1.Equal(creationTime2) {
+		return core.GetNamedspacedName(podInfo1.Pod) < core.GetNamedspacedName(podInfo2.Pod)
 	}
-	return pgInfo, podMinAvailable
+	return creationTime1.Before(creationTime2)
 }
 
 // PreFilter performs the following validations.
-// 1. Validate if minAvailables and priorities of all the pods in a PodGroup are the same.
-// 2. Validate if the total number of pods belonging to the same `PodGroup` is less than `minAvailable`.
+// 1. Validate if minMember and priorities of all the pods in a PodGroup are the same.
+// 2. Validate if the total number of pods belonging to the same `PodGroup` is less than `minMember`.
 //    If so, the scheduling process will be interrupted directly to avoid the partial Pods and hold the system resources
 //    until a timeout. It will reduce the overall scheduling time for the whole group.
+// 3. Validate if resource is enough, if not enough, pod would not pass this check, and the the group would be ad to a denyList
 func (cs *Coscheduling) PreFilter(ctx context.Context, state *framework.CycleState, pod *v1.Pod) *framework.Status {
-	pgInfo, podMinAvailable := cs.getOrCreatePodGroupInfo(pod, time.Now())
-	pgKey := pgInfo.key
-	if len(pgKey) == 0 {
-		return framework.NewStatus(framework.Success, "")
-	}
-	pgMinAvailable := pgInfo.minAvailable
-
-	// Check if the values of minAvailable are the same.
-	if podMinAvailable != pgMinAvailable {
-		klog.V(3).Infof("Pod %v has a different minAvailable (%v) as the PodGroup %v (%v)", pod.Name, podMinAvailable, pgKey, pgMinAvailable)
-		return framework.NewStatus(framework.Unschedulable, "PodGroupMinAvailables do not match")
-	}
-	// Check if the priorities are the same.
-	pgPriority := pgInfo.priority
-	podPriority := podutil.GetPodPriority(pod)
-	if pgPriority != podPriority {
-		klog.V(3).Infof("Pod %v has a different priority (%v) as the PodGroup %v (%v)", pod.Name, podPriority, pgKey, pgPriority)
-		return framework.NewStatus(framework.Unschedulable, "Priorities do not match")
-	}
-
-	total := cs.calculateTotalPods(pgInfo.name, pod.Namespace)
-	if total < pgMinAvailable {
-		klog.V(3).Infof("The count of PodGroup %v (%v) is less than minAvailable(%d) in PreFilter: %d",
-			pgKey, pod.Name, pgMinAvailable, total)
-		return framework.NewStatus(framework.Unschedulable, "less than pgMinAvailable")
+	if err := cs.pgMgr.PreFilter(ctx, pod); err != nil {
```

Comment:
Why we move the logic to the PodGroupManager?
Keep the implement of coscheduling in coscheudling.go and just move the function related to podgroup to pg.go